benchmark/benchmarks.jl

```diff
@@ -43,3 +51,305 @@ SUITE["model"]["initialise_union"]["continuous"] = @benchmarkable ABM(
     warn = false,
 )
 
+#### API ###
+
+for space in ["graph", "grid", "continuous"]
+    SUITE[space] = BenchmarkGroup(["add", "add_union", "move", "neighbors"])
+
+    SUITE[space]["add"] = BenchmarkGroup([
+        "agent",
+        "agent_pos",
+        "agent_single",
+        "create",
+        "create_pos",
+        "create_single",
+    ])
+    SUITE[space]["add_union"] = BenchmarkGroup(["agent", "agent_pos", "agent_single"])
+    if space == "continuous"
+        SUITE[space]["move"] = BenchmarkGroup(["update", "vel"])
+        SUITE[space]["neighbors"] = BenchmarkGroup([
+            "nearby_ids",
+            "nearby_agents",
+            "nearby_ids_iterate",
+            "nearby_agents_iterate",
+            "nearest",
+            "interacting",
+        ])
+    else
+        SUITE[space]["move"] = BenchmarkGroup(["random", "pos", "single"])
+        SUITE[space]["neighbors"] = BenchmarkGroup([
+            "nearby_ids",
+            "nearby_agents",
+            "nearby_ids_iterate",
+            "nearby_agents_iterate",
+            "position_pos",
+            "position_agent",
+        ])
+    end
+    SUITE[space]["collect"] = BenchmarkGroup(["init_agent", "store_agent"])
+end
+# some spaces have specific things we'd like to add
+push!(SUITE["grid"]["add_union"].tags, "agent_fill")
+push!(SUITE["grid"]["add"].tags, "create_fill")
+for space in ["grid", "graph"]
+    push!(SUITE[space].tags, "position")
+    SUITE[space]["position"] = BenchmarkGroup(["contents", "agents"])
+end
+# some spaces need a few things dropped
+for add in ["add", "add_union"]
+    group = SUITE["continuous"][add].tags
+    add == "add" && splice!(group, findfirst(t -> t == "create_single", group))
+    splice!(group, findfirst(t -> t == "agent_single", group))
+end
+
+function iterate_over_neighbors(a, model, r)
+    s = 0
+    for x in nearby_ids(a, model, r)
+        s += x
+    end
+    return s
+end
+
+#### API -> GRAPH ####
+
+graph_model = ABM(GraphAgent, GraphSpace(complete_digraph(200)))
+graph_agent = GraphAgent(1, 82, 6.5, false)
+graph_union_model = ABM(
+    Union{GraphAgent,GraphAgentTwo,GraphAgentThree,GraphAgentFour,GraphAgentFive},
+    GraphSpace(complete_digraph(200)),
+    warn = false,
+)
+
+# Limit samples here so space does not saturate with agents
+SUITE["graph"]["add"]["agent"] =
+    @benchmarkable add_agent!($graph_agent, $graph_model) samples = 100
+SUITE["graph"]["add"]["agent_pos"] =
+    @benchmarkable add_agent_pos!($graph_agent, $graph_model) samples = 100
+SUITE["graph"]["add"]["agent_single"] =
+    @benchmarkable add_agent_single!($graph_agent, $graph_model) samples = 100
+SUITE["graph"]["add"]["create_pos"] =
+    @benchmarkable add_agent!(26, $graph_model, 6.5, false) samples = 100
+SUITE["graph"]["add"]["create_single"] =
+    @benchmarkable add_agent_single!($graph_model, 6.5, false) samples = 100
+SUITE["graph"]["add"]["create"] =
+    @benchmarkable add_agent!($graph_model, 6.5, false) samples = 100
```

Comment:
Notice that you benchmark all types of add in all spaces. Yet, only the `add_agent_pos` is different between different spaces, all other add types are centralized in the API.jl file. Since the other functions are in fact the same and should be timed in only one space (GraphSpace for simplicity), to reduce time and complexity of output

Reply:
Fair points on both. I'll alter things once I manage to get it running correctly.